handler/src/main/java/io/netty/handler/ssl/SniHandler.java

```diff
@@ -307,20 +307,39 @@ public void operationComplete(Future<SslContext> future) throws Exception {
 
     private void replaceHandler(ChannelHandlerContext ctx, Selection selection) {
         SSLEngine sslEngine = null;
+        SslHandler handler = null;
         this.selection = selection;
         try {
             sslEngine = selection.context.newEngine(ctx.alloc());
-            ctx.pipeline().replace(this, SslHandler.class.getName(), selection.context.newHandler(sslEngine));
+            handler = SslContext.newHandler(sslEngine);
+            replaceHandler(ctx, selection.hostname, selection.context, SslHandler.class.getName(), handler);
         } catch (Throwable cause) {
             this.selection = EMPTY_SELECTION;
             // Since the SslHandler was not inserted into the pipeline the ownership of the SSLEngine was not
             // transferred to the SslHandler.
             // See https://github.com/netty/netty/issues/5678
             ReferenceCountUtil.safeRelease(sslEngine);
+
+            if (handler != null) {
+                handler.sslCloseFuture().cancel(true);
+            }
             ctx.fireExceptionCaught(cause);
         }
     }
 
+    /**
+     * Users may override this method to implement custom behavior. The default
+     * implementation will simply replace {@code this} {@link SniHandler} instance
+     * with the given {@link SslHandler}.
+     *
+     * BEWARE: The {@code hostname} argument can be {@code null}.
+     */
+    protected void replaceHandler(ChannelHandlerContext ctx, String hostname,
```

Comment:
Looking at the internals of SslHandler the earlier mentioned `sslCloseFuture()` I'm wondering if this method shouldn't be something like this instead...

```java
replaceHandler(ChannelHandlerContext ctx, String hostname, SslContext context, SSLEngine engine) {
  ctx.pipeline().replace(this, SslHandler.class.getName(), new SslHandler(engine));
}
```

This would allow users to extend SslHandler as well and I don't think `sslCloseFuture()` is wall suited for rer counting right now (not sure it'll always fire and in the expected order in respect to SSLEngine).